pkg/identity/identity.go

```diff
@@ -180,18 +181,33 @@ func (pair *IPIdentityPair) PrefixString() string {
 // RequiresGlobalIdentity returns true if the label combination requires a
 // global identity
 func RequiresGlobalIdentity(lbls labels.Labels) bool {
-	needsGlobal := true
+	return ScopeForLabels(lbls) == IdentityScopeGlobal
+}
+
+// ScopeForLabels returns the identity scope to be used for the label set.
+// If all labels are either CIDR or reserved, then returns the CIDR scope.
+// Note: This assumes the caller has already called LookupReservedIdentityByLabels;
+// it does not handle that case.
+func ScopeForLabels(lbls labels.Labels) NumericIdentity {
+	scope := IdentityScopeGlobal
+
+	// If this is a remote node, return the remote node scope.
+	// Note that this is not reachable when policy-cidr-selects-nodes is false, since
+	// callers will already have gotten a value from LookupReservedIdentityByLabels.
+	if lbls.Has(labels.LabelRemoteNode[labels.IDNameRemoteNode]) {
+		return IdentityScopeRemoteNode
+	}
 
 	for _, label := range lbls {
 		switch label.Source {
 		case labels.LabelSourceCIDR, labels.LabelSourceReserved:
-			needsGlobal = false
+			scope = IdentityScopeLocalCIDR
```

Comment:
Seems a bit odd to define `LabelSourceReserved` as requiring CIDR scope, since reserved identities aren't allocated in the local CIDR range. 

I'm not sure the implications of this, but at the very least we should probably add a comment here that reserved identities are just being treated like local CIDR scope because the semantics are closer to LocalCIDR than they are to ScopeGlobal.

Reply:
It seems wrong that all reserved identities would get lumped into the CIDR scope? Maybe a separate non-CIDR local scope is needed for them?

Reply:
Right; none of the behavior changes, it's just naming. I don't know of any other "reserved" identities that are local besides CIDR identities, but I agree the naming is poor.

How about ScopeLocal and ScopeRemoteNode? That at least makes some sense. (Local is confusing, it means local identities, not local destinations, but comments should clear that up).

Reply:
I think this is where the desired behaviours don't 100% line up with the bitpatterns used in the datapath, since reserved identities are be `0x000000xx` (scope: global) but they should be treated more like local identities (`0x01xxxxxx`). This function only checks for presence of the reserved label, and we could actually allocate `0x01xxxxxx` identities for these label sets particularly if they also include CIDR labels.

Put another way, there are:
1. Reserved identities in the range 1-255 with only `reserved:foo` labels
2. Local identities in the range `0x01xxxxxx` with CIDR labels and without `reserved:foo` labels
3. Local identities in the range `0x01xxxxxx` with CIDR labels and with `reserved:foo` labels (new in this PR)

The previous version of this function was answering the question "Does this identity need global identity allocation?" and the answer is the same for all three. I wonder about whether (1) and (3) belong to the same "scope" or whether perhaps (1) should be its own special scope. But in practical terms, I don't think that any code is making the wrong assumption based on the new version of this code introduced in this PR.